### DISCUSSION TITLE
DataStore Foreign Keys

### DISCUSSION THREAD

**Discussion by JVickery-TBS** [ckan/ckan#8542] [Ideas]
Related fork PR: https://github.com/open-data/ckan/pull/181

Support Foreign Keys in DataStore. Above is a rudimentary version of it which is limited by another plugin to a certain Dataset type which is controlled by sysadmin/system things so no issues there.

But actually implementing this, and supporting this would require a lot more. I would like to discuss how to handle Foreign Key constraints on deletion and updates. Does a user have to have permissions on a DataStore Resource to reference it? Or do we make a whole 'nother implement like a "Privacy Settings" for DataStore resources, which would allow your DataStore resource to be referenced by another?

Like, by default a DataStore resource is "private" so you cannot reference it in a Foreign Key constraint. But a schema and view for Resources could do "Public Reference" and then also set "Cascade on Update", "Cascade on Delete", "Constraint on Update", "Constraint on Delete", etc etc.

The main problem I have is with permissions here. How should we handle the permissions for Foreign Keys, because I don't think we want to limit an owner's DataStore resource from another user referencing it.

**Comment by wardi** [comment]
Here are the constraints we should consider:

- MATCH FULL
  should be our default because the postgres "MATCH SIMPLE" default lets you set some columns in a FK to NULL to allow the to FK not match anything

  (this only affects composite foreign keys)
- ON DELETE SET NULL
  should be our default because it's the least destructive and restrictive option, but will act like NO ACTION if the column is NOT NULL

  we could prevent normal users from applying a NOT NULL check on columns that are part of a FK to avoid the issue.
- ON DELETE CASCADE
  should be available with a warning that all data in the row (or the entire table) will be removed when parent rows are.
- ON DELETE NO ACTION
  essentially the same as RESTRICT, should be available but only when the owner of the parent table has volunteered the table to be restricted this way by other users

  how we actually record volunteered tables is another question. It's a property of the table and we don't have a place to store those. We could add a comment on the table like we do for columns and the data dictionary, but something like https://github.com/ckan/ckan/discussions/8531 might be more elegant
- ON UPDATE CASCADE
  should be our default so that FKs stay in sync with their referenced column
- ON UPDATE NO ACTION
  essentially the same as RESTRICT and again, the parent table would need to volunteer to be restricted this way.

So, for an initial implementation without the "volunteer table to be restricted" part we could only support foreign keys with:

    MATCH FULL ON DELETE SET NULL ON UPDATE CASCADE

and disallow NOT NULL on the FK columns.

**Comment by wardi** [comment]
We should also make sure there are indexes on the FK fields so that performance of updating the parent tables isn't significantly affected by the foreign keys from other tables.
